### DISCUSSION TITLE
question concerning api key expiration and websockets

### DISCUSSION THREAD

**Discussion by hadfield** [deepgram/community#673] [General help]
I'm planning the processes around generating api keys to use on the client side of a web application that will involve live streaming of audio for transcription.

What happens if an api key expires while a websocket is open? (either idle or sending audio.)

I'm looking for a way to ensure that an expired api key is not being used to transcribe audio.

Ideally deepgram would force close the websocket upon such an api key expiration.

Please advise!

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by jjmaldonis** [comment]
Hey @hadfield, if the API key expires while the websocket connection is open, the connection will stay open. Authentication is performed when the websocket connection is initialized, similar to an HTTP request. This is standard practice for internet communication protocols.

There are a few ways you could implement the functionality you're looking for in your app.
1. You could implement logic to periodically check the status of the API key, and check whether or not the API key is expired. If the API key is expired, you could disconnect the websocket.
2. You could rely on API key expirations by periodically closing the websocket and reopening it. This may work very well if you have a ~10-15 minute tolerance for keeping the websocket open after the API key expires because you could reconnect the websocket every 10-15 minutes.
3. You could implement a different authentication mechanism between your client and your backend, and disconnect the client-side websocket depending on the response from the backend.

**Reply by hadfield** [reply]
Thanks for the comment!

From a security perspective, I'm trying to account for the case of an API Key being compromised and being able to expire the key immediately and prevent any additional resources being consumed via the key.  If the application client code periodically closes and re-opens the websocket that wouldn't prevent "attacker" client code from simply not doing that.

Certainly I could have a proxy websocket server in the middle, but then would need to stand up and support this extra infrastructure.

If on the deepgram side the websocket enforced re-authenticating every N hours then at least a compromised key would have an enforced expiration time rather than being able to be used indefinitely as long as the connection was held open.

**Comment by eddie-hopeharbor** [comment]
I have a "2-minute translation free trial" feature for unregistered users. On browser, it requests a 2-minute TTL temporary API key from the backend and opens up the websocket connection directly with deepgram. 
A hacker can, in theory, write a client-side script, such that after the websocket connection with deepgram is established, it blocks any HTTP requests from the browser to my backend server. It also blocks/interferes my client code that regularly checks the key expiration and closes the websocket. At this point my backend server is totally in the dark without knowing whether or not the browser is using deepgram indefinitely behind my back.
Worst case, the hacker publishes and distributes the script on userscript platforms like violentmonkey for free or for a fee. Then a large number people can freely use my service in trial mode for free indefinitely.
These are unlikely to happen, but in theory it can. Ideally deepgram should throw error in the middle of a connection if the key expires, and provide a way for us to feed a new API key before the old one expires, all without closing and reopening the connection. Microsoft Azure real-time translation API/websocket does exactly this for client side usage.

**Reply by davidvonthenen** [reply]
Hi @eddie-hopeharbor 

My advice would be to implement Jason's option 3 for production applications. In my opinion, the temporary API Keys are just too problematic, and any security-conscious person would never use them for exactly the reasons you pointed out. For me, the temporary keys are great for quick and dirty demos and the "it's a lot of effort to implement option 3" but at the cost of security.

Just my 2 cents.

**Reply by hedgeh0k** [reply]
@eddie-hopeharbor could you please clarify how to set expiration when creating a temporary API token? I see no such a param for this in docs https://developers.deepgram.com/docs/create-additional-api-keys

**Comment by hadfield** [comment]
Re-opening as the fundamental issue is not solved.

**Comment by SandraRodgers** [comment]
Hi @nikita-mityagin and @hadfield 

Here is a Deepgram blog post that gives information about how to set the expiration when creating a temporary API key. See approaches 1 and 2: https://deepgram.com/learn/protecting-api-key

And here's a code example in node which uses approach 2: https://github.com/deepgram-devs/deepgram-server/blob/main/server.js#L31

I hope this is the information that you're looking for!

**Reply by hedgeh0k** [reply]
Sandra, thank you!

**Reply by hadfield** [reply]
Does this now have the behavior as described by @eddie-hopeharbor above:

> Ideally deepgram should throw error in the middle of a connection if the key expires, and provide a way for us to feed a new API key before the old one expires, all without closing and reopening the connection. Microsoft Azure real-time translation API/websocket does exactly this for client side usage.

?

**Reply by jhartquist** [reply]
@SandraRodgers @hadfield 
No. Once the socket is open, it remains open even after TTL/expiry. I think @davidvonthenen is correct in that creating a proxy is the most secure way to protect against this attack.
